rfcs/0004-checkpoints.md

```diff
@@ -0,0 +1,516 @@
+current set of L0s and Sorted Runs are synchronized between the Writer and Compactor through updates to the centralized
+manifest. When reading, the Writer gets a reference to a copy of the metadata for the current set of sorted runs and
+searches them for the requested data. Meanwhile, the GC scans the SSTs in Object Storage and deletes any SSTs that are
+not referenced by the current manifest and are older than some threshold.
+
+## Problem
+
+The current implementation poses a few problems and limitations:
+1. There is nothing preventing the GC from deleting SSTs out from the Writer while it’s attempting to read their data.
+   This isn’t a big deal today as we only support key lookups, which are fast. But this could become a more acute
+   problem when we support range scans which can be long.
+1. Currently only the Writer process can interact with the database. We’d like to support additional read-only clients,
+   called Readers that can read data.
+1. Some use cases can benefit from forking the database and establishing a new db that relies on the original DB’s SSTs
+   but writes new data to a different set of SSTs from the original writer.
+
+## Goals
+
+In this RFC, we propose adding checkpoints and clones to address this limitation.
+- Clients will be able to establish cheap db checkpoints that reference the current version of SlateDB’s manifest.
+- Writers will use checkpoints to “pin” a version of the database for reads to avoid the GC from deleting the SSTs.
+- Readers will use checkpoints similarly, and many Reader processes can coexist alongside a Writer.
+- Finally, we will support creating so-called “clones” from a checkpoint to allow spawning new Writers that write to
+  a “fork” of the original database.
+
+### Out Of Scope
+
+This RFC does not propose:
+- Any mechanism to provide isolation guarantees from other concurrent reads/writes to the database.
+
+## Proposal
+
+At a high level, this proposal covers a mechanism for establishing database checkpoints, and reviews how the various
+processes will interact with this mechanism. Logically, a checkpoint refers to a consistent and durable view of the
+database at some point in time. It is consistent in the sense that data read from a checkpoint reflects all writes up
+to some point in time, and durable in the sense that this property holds across process restarts. Physically, a
+checkpoint refers to a version of SlateDB’s manifest.
+
+### Manifest Schema
+
+Checkpoints themselves are also stored in the manifest. We’ll store them using the following schema:
+
+```
+
+table DbParent {
+   // Optional path to a parent database from which this database was cloned
+   parent_path: string (required);
+
+   // Optional parent checkpoint ID
+   parent_checkpoint: UUID (required);
+}
+
+table ManifestV1 {
+   // Optional details about the parent checkpoint for the database
+   parent: DbParent
```

Comment:
At some point, the fact that this db was cloned from another wouldn't matter. i.e. All the data would be new data and is written in this DB. 

Would this field ever be set to null, after it had an initial value?